### PR TITLE
Fan curve support

### PR DESCRIPTION
**Pull request by JackDoan**
Hello!

This isn't quite ready yet, but I wanted to share it with you and get your thoughts on my approach.

### CODE REVIEW

**Review by aleksamagicka — CHANGES_REQUESTED** [review]
Excellent work, thank you so much! I haven't yet tested this on my pump, so my comments are coming from reading the code. I'll test it in the upcoming days. Comments mostly tackle what I've seen the maintainers say

File: aquacomputer_d5next.c
Comment (JackDoan) [reply aquacomputer_d5next.c]: Done!

File: aquacomputer_d5next.c
Comment (aleksamagicka) [reply aquacomputer_d5next.c:568]: TIL about MISRA. This is in no way that critical 😃

File: aquacomputer_d5next.c
Comment (JackDoan) [reply aquacomputer_d5next.c:439]: Yeah, I spent ages looking and couldn't find anything sane. The preprocessor doesn't really do loops.

File: aquacomputer_d5next.c
Comment (JackDoan) [reply aquacomputer_d5next.c]: Following the example of it87 and nct6775, if curve support exists it is always shown. This is more for hiding features the current HW doesn't support.

Otherwise, you would be unable to configure the fan curve before enabling it, which would be bad.

File: aquacomputer_d5next.c
Comment (JackDoan) [reply aquacomputer_d5next.c:570]: Also, we do go straight to unlock, there's an `else` branch that gets skipped. Is that okay, or does it stylistically make more sense to drop the `else` and use `goto`?

File: aquacomputer_d5next.c
Comment (JackDoan) [reply aquacomputer_d5next.c:323]: Are you sure? I wasn't wild about a switch with 16 identical branches.

File: aquacomputer_d5next.c
Comment (JackDoan) [reply aquacomputer_d5next.c:177]: Heh, when I see defines, I usually think "could these be enums?"

File: aquacomputer_d5next.c
Comment (aleksamagicka) [reply aquacomputer_d5next.c:323]: I'm only saying that because most of the hwmon code I saw was using switches for this type of stuff. If statements are generally fine by me, just saying it could be a switch since it depends on a single variable

File: aquacomputer_d5next.c
Comment (JackDoan) [reply aquacomputer_d5next.c:305]: >Also add a note above what it does because of the additional * 100
I'm not totally sure what you're referring to.

>Should be d5next_percent_to_pwm
Yeah, that's clearer. Good call.

File: aquacomputer_d5next.c
Comment (aleksamagicka) [reply aquacomputer_d5next.c:571]: I'd say the latter. Most of hwmon (and probably the rest of kernel) code jumps out as soon as an error occurs. Look at `d5next_probe` for example, I copied it from other drivers

File: aquacomputer_d5next.c
Comment (aleksamagicka) [reply aquacomputer_d5next.c]: For the *100, I thought about adding a note saying that the pump expects the value to be multiplied by 100, so that's why it's done. Sorry about being unclear

File: aquacomputer_d5next.c
Comment (aleksamagicka) [reply aquacomputer_d5next.c]: Oh, so auto refers to the curves? If so, don't hide it as you say

**Comment by JackDoan** [comment]
Broadly speaking, how do you feel about changing the order of the pump and fan? The fan comes first in the control packet, so it makes more sense to me to have the fan be  pwm/fan/current/in_1, _but_, since this is already in mainline, I worry that renumbering things could cause trouble for existing users.

This branch currently has pwm/fan 1 and 2 swapped from how they are on mainline (fan = 1, pump  = 2). Let me know your thoughts and I can put it back the way I found it -- it was just marginally more convenient to write the code this way.

File: aquacomputer_d5next.c
Comment (JackDoan) [reply aquacomputer_d5next.c:332]: I use `exit:` in one other place, but there it isn't unlocking, it really just is exiting.

**Comment by aleksamagicka** [comment]
I think it's best to make it like it already is in mainline. I'm afraid changing the order would possibly fall into "we don't break userspace ever".

File: aquacomputer_d5next.c
Comment (JackDoan) [reply aquacomputer_d5next.c]: I wasn't wild about adding these to the docs until I actually made the driver use them. I'd throw them into re_docs, but that's not on my branch yet, I need to squash & rebase.

File: aquacomputer_d5next.c
Comment (JackDoan) [reply aquacomputer_d5next.c]: That's correct. Specifically, it refers to fan control mode `2`. Mode `1` is PID, which has a different suite of settings that I haven't exposed yet. (not sure if I want to. Hwmon doesn't seem to have an idiomatic way to expose PID tuning params :P)

**Comment by JackDoan** [comment]
Understood! I'll make the needed adjustments.

File: aquacomputer_d5next.c
Comment (aleksamagicka) [reply aquacomputer_d5next.c]: Yeah, if it can't use an existing interface it's best left to userspace. Similar story with RGB.

**Comment by aleksamagicka** [comment]
Thanks! I'll take a look & test as soon as I can.

File: aquacomputer_d5next.c
Comment (aleksamagicka) [inline aquacomputer_d5next.c:567]: Probably shouldn't be using macros for this. In the [docs](https://www.kernel.org/doc/html/latest/hwmon/submitting-patches.html), it says: `Avoid macros to generate groups of sensor attributes. It not only confuses checkpatch, but also makes it more difficult to review the code.`. I looked at some other drivers randomly and they have use `SENSOR_DEVICE_ATTR_2` directly.

**Review by aleksamagicka — COMMENTED** [review]
I tested it. Reading pwm1 and pwm2 returns 0 always, even though the pump and fan are running. Writing to the curve controls crashes the driver. Sorry for not investigating further, I'm a bit tight on time right now 😅  Just wanted to let you know what I found. I'm on firmware 1021, maybe something changed?

File: aquacomputer_d5next.c
Comment (JackDoan) [reply aquacomputer_d5next.c:567]: Oof the crashes are embarrassing! I should write some integration tests. I'm on fw 1019.

Good to know, I'll unroll the macros.

Comment (JackDoan) [reply aquacomputer_d5next.c:566]: >Reading pwm1 and pwm2 returns 0 always

When you say this, do you mean in all modes, or just curve mode? They work fine for me in manual mode. Testing curve mode now...

Comment (JackDoan) [reply aquacomputer_d5next.c:566]: Hmm, curve mode behaves as expected for me too on `pwm2`. I'm going to unroll those macros, and investigate potential FW differences.

Comment (aleksamagicka) [reply aquacomputer_d5next.c:566]: Curve mode

Comment (JackDoan) [reply aquacomputer_d5next.c]: Would you mind showing me the output of `cat pwm2_*` after you enable curve mode with `echo 2 > pwm2_enable`?

The value of `temp1_input` would be helpful as well :)

Comment (aleksamagicka) [reply aquacomputer_d5next.c]: [Here](https://gist.github.com/aleksamagicka/f8710dfc580be77a2b0a7fb839a14554) it is. Water temp is `30670`. Setting curve points doesn't crash it now :+1:

Comment (JackDoan) [reply aquacomputer_d5next.c]: That's interesting, because I'm not sure I actually did anything to fix the "bug". :/

Comment (aleksamagicka) [reply aquacomputer_d5next.c]: OK, I'll retest soon. I also have news from the kernel side. Recently I've sent in the Farbwerk 360 driver I've written in the meanwhile and Guenter (the hwmon maintainer) requested that it be incorporated into the existing one, so I did that and that's live in hwmon-next. So when this is completed, we'll have to rebase this (so to say) on top of that. Because of that I'll try to get the Octo driver out the door ASAP to get his feedback and minimize the final changes needed here, as it will have the basic PWM write support (so getting and sending reports).

Comment (JackDoan) [reply aquacomputer_d5next.c]: Are you planning to merge the octo driver into the d5/farbwerk driver as well?

Comment (aleksamagicka) [reply aquacomputer_d5next.c]: Yes, hopefully I'll send it tomorrow. Then, once it gets merged, I'll work on combining yours and that (will probably open a PR at your repo, or push here somehow if that's possible) so we can continue where we left off (I didn't really plan he'll request them to be combined, but it makes sense in hindsight)

**Pull request closed without merging**
